design/kata-2-0-metrics.md

```diff
@@ -0,0 +1,1343 @@
+
+- Get metrics
+- Get events
+- Extend hypervisor commands
+
+In this document we will cover metrics only. And until now it only supports metrics function.
+
+This is the architecture overview metrics in Kata Containers 2.0.
+
+![Kata Containers 2.0 metrics](arch-images/kata-2-metrics.png)
+
+
+And the sequence diagram is shown below:
+
+![Kata Containers 2.0 metrics ](arch-images/kata-metrics-sequence-diagram.png)
+
+
+### Kata magent
+
+`magent` is a management agent on one nodes, where many Kata containers are running. `magent`'s work include:
+
+- Aggregate sandbox metrics running on this node, and add `sandbox_id` label
+- As a prometheus target, all metrics from Kata shim on this node will be collected by prometheus indirectly. This can easy the targets count in Prometheus, and also need not to expose shim's metrics by `ip:port`
+
+Only one `magent` process are running on one node.
+
+`magent` is using a different communication channel other than that `conatinerd` communicating with Kata shim, and Kata shim listen on a new socket address for communicating with `magent`.
+
+The way `magent` get shim's metrics socket file(`magent_address`) like that `containerd` get shim address. The socket is an abstract socket and saved as file `abstract` with the same directory of `address` for `containerd`.
+
+**Note**: If there is no Prometheus server is configurated, i.e., there is no scrape operations, `magent` will do nothing initiative.
+
+### Kata runtime
+
+Runtime are mainly to:
+
+- Gather metrics about shim process
+- Gather metrics about hypervisor process
+- Gather metrics about running sandbox
+- Get metrics from Kata agent(through `ttrpc`)
+
+### Kata agent
+
+Agent is responsible for:
+
+- Gather agent process metrics
+- Gather guest OS metrics
+
+And in Kata 2.0, agent will add a new interface:
+
+```
+rpc GetMetrics(GetMetricsRequest) returns (Metrics);
+
+message GetMetricsRequest {}
+
+message Metrics {
+	string metrics = 1;
+}
+
+```
+
+The `metrics` field is Prometheus encoded content. This can avoid defining a fixed struct in protocol buffers.
+
+### Performance and overhead
+
+Metrics should not become the bottleneck of system, downgrade the performance, and run with minimal overhead.
+
+|\*|No Sandbox | 1 Sandbox | 2 Sandboxes |
+|---|---|---|---|
+|Metrics count| 39 | 106 | 173 |
+|Metrics size(bytes)| 9K | 144K | 283K |
+|Metrics size(gzipped, bytes)| 2K | 10K | 17K |
+
+*Metrice size*: Response size of one Prometheus scrape request.
+
+It's easy to estimated that if there are 10 sandboxes running in the host, the size of one metrics fetch request issued by Prometheus will be about to 9 + (144 - 9) * 10 = 1.35M (un-gzipped) or 2 + (10 - 2) * 10 = 82K (gzipped). Of course Prometheus support gzip compression, that can reduce the response size of every request.
+
+And here is some test data:
+
+- E2E(from request to `magent` and `magent` write response back): 20ms(avg)
```

Comment:
You may find "E2E" fails the spell check as it's not a word the checker knows about. I'd suggest writing this out in full:

> End-to-end ("E2E")

... but if you want to reuse the term in other parts of the document also consider adding it to the custom dictionary:

- https://github.com/kata-containers/tests/tree/master/cmd/check-spelling#adding-a-new-word